TPP.Core/Commands/Definitions/BadgeCommands.cs

```diff
@@ -271,15 +277,33 @@ public async Task<CommandResult> Pokedex(CommandContext context)
         public async Task<CommandResult> GiftBadge(CommandContext context)
         {
             User gifter = context.Message.User;
-            (User recipient, PkmnSpecies species, Optional<PositiveInt> amountOpt) =
-                await context.ParseArgs<AnyOrder<User, PkmnSpecies, Optional<PositiveInt>>>();
+            (User recipient, PkmnSpecies species, Optional<PositiveInt> amountOpt, Optional<Badge.BadgeSource> sourceOpt, Optional<Shiny> shinyOpt, Optional<string> formOpt) =
+                await context.ParseArgs<AnyOrder<User, PkmnSpecies, Optional<PositiveInt>, Optional<Badge.BadgeSource>, Optional<Shiny>, Optional<string>>>();
             int amount = amountOpt.Map(i => i.Number).OrElse(1);
-
+            Badge.BadgeSource? source = sourceOpt.IsPresent ? sourceOpt.Value : null;
+            int? form = null;
+            bool? shiny = shinyOpt.IsPresent ? shinyOpt.Value : false;
+            if (formOpt.IsPresent)
+            {
+                string formName = formOpt.Value;
+                try
+                {
+                    form = PkmnForms.getFormId(species, formName);
+                }
+                catch (ArgumentException e)
+                {
+                    return new CommandResult
+                    {
+                        Response = e.Message
+                    };
+                }
+            }
```

Comment:
This large block of code is duplicated a few times. You can either extract it and/or reduce its size. I personally find the following quite aesthetically acceptable:
```c#
int? form;
try {
    form = formOpt
        .Map(name => PkmnForms.getFormId(species, name))
        .OrElse(null);
} catch (ArgumentException e) {
    return new CommandResult { Response = e.Message };
}
```

Reply:
This was originally much worse, and will almost certainly be moved to a method along with other related badge info.